[SYCL-MLIR] Implement `OwnerLessBase` SYCL type

whitneywhtsang opened the pull request:
Please only review https://github.com/intel/llvm/pull/7648/commits/e1c7f35ece8792eb1447959537612caeaaea3d1e.

d9484274cd1a21f90509cb79f0cd5f425f8f9e3b was reverted in 9f5ba51229b090fc6af1db1b7b25494d241f86b1. This PR adds back d9484274cd1a21f90509cb79f0cd5f425f8f9e3b.
d9484274cd1a21f90509cb79f0cd5f425f8f9e3b adds `OwnerLessBase` as base class of `accessor` type.

This PR implements the SYCL type `OwnerLessBase`.
clang generates `%"class.sycl::_V1::detail::OwnerLessBase" = type { i8 }`.

Please do not squash and merge this PR.

etiotto commented:
Reviewed https://github.com/intel/llvm/commit/2deae0d505f65ad241e54df7334deb936dab5a8d. LGTM (added some comments inline).

whitneywhtsang commented:
> Reviewed [2deae0d](https://github.com/intel/llvm/commit/2deae0d505f65ad241e54df7334deb936dab5a8d). LGTM (added some comments inline).

Addressed all review comments.

whitneywhtsang commented:
There are merge conflicts when merging from `sycl` branch, due to missing d9484274cd1a21f90509cb79f0cd5f425f8f9e3b in `sycl-mlir` branch. Merging this PR without 2 approvals to unblock the merge. Feel free to give post commit review comments if needed.
